src/lib.rs

```diff
@@ -1,41 +1,385 @@
+                && existing_rectangle.lower_right_corner.y > rectangle.upper_left_corner.y
+        }) {
+            panic!("Given NFT rectangle collides with already minted one");
+        }
+
+        // Painting
+
+        let rectangle_width = rectangle.width() as usize;
+        let rectangle_height = rectangle.height() as usize;
+        let rectangle_pixel_count = get_pixel_count(rectangle_width, rectangle_height);
+
+        check_painting(&painting, rectangle_pixel_count);
+        paint(
+            self.resolution,
+            &rectangle,
+            rectangle_width,
+            rectangle_height,
+            &mut self.painting,
+            painting,
+        );
+
+        // Payment and NFT minting
+
+        transfer_ftokens(
+            self.ft_program,
+            msg::source(),
+            self.owner,
+            rectangle_pixel_count as u128 * self.pixel_price,
+        )
+        .await;
+
+        let token_id = mint_nft(self.nft_program, token_metadata).await;
+        transfer_nft(self.nft_program, msg::source(), token_id).await;
+
+        // Insertion and replying
+
+        self.tokens_by_rectangles.insert(
+            rectangle,
+            TokenInfo {
+                owner: msg::source(),
+                pixel_price: None,
+                token_id,
+            },
+        );
+        self.rectangles_by_token_ids.insert(token_id, rectangle);
+
+        reply(NFTPixelboardEvent::Minted(token_id));
+    }
 
-    if payload == "Hello" {
-        msg::reply(b"World", 0).unwrap();
+    async fn buy(&mut self, token_id: TokenId) {
+        let (rectangle, token) = get_mut_token_by_id(
+            &self.rectangles_by_token_ids,
+            &mut self.tokens_by_rectangles,
+            token_id,
+        );
+
+        let pixel_price = token
+            .pixel_price
+            .unwrap_or_else(|| panic!("NFT isn't for sale"));
+        // get_pixel_count() isn't used here because it checks an NFT area for
+        // equality to 0, but here it's always not equal 0.
+        let token_price =
+            (rectangle.width() as usize * rectangle.height() as usize) as u128 * pixel_price;
+        let resale_commision = token_price * self.commission_percentage as u128 / 100;
+
+        transfer_ftokens(self.ft_program, msg::source(), self.owner, resale_commision).await;
+        transfer_ftokens(
+            self.ft_program,
+            msg::source(),
+            token.owner,
+            token_price - resale_commision,
+        )
+        .await;
+
+        token.pixel_price = None;
+        token.owner = msg::source();
+        transfer_nft(self.nft_program, msg::source(), token_id).await;
+
+        reply(NFTPixelboardEvent::Bought(token_id));
+    }
+
+    async fn put_up_for_sale(&mut self, token_id: TokenId, pixel_price: u128) {
+        let token = get_mut_token_by_id(
+            &self.rectangles_by_token_ids,
+            &mut self.tokens_by_rectangles,
+            token_id,
+        )
+        .1;
```

Comment:
Maybe it is more readable but up to you.
```suggestion
        let (_, token) = get_mut_token_by_id(
            &self.rectangles_by_token_ids,
            &mut self.tokens_by_rectangles,
            token_id,
        );
```